web-runtime/src/marine_js.rs

```diff
@@ -0,0 +1,305 @@
+    }
+
+    pub fn exports(&self) -> ExportIter {
+        ExportIter::new(&self.exports)
+    }
+}
+
+pub struct DynFunc<'a> {
+    pub(crate) signature: FuncSig,
+    pub name: String,
+    pub module_name: String,
+    //pub(crate) instance_inner: &'a InstanceInner,
+    //func_index: FuncIndex,
+    data3: PhantomData<&'a i32>,
+}
+
+impl<'a> DynFunc<'_> {
+    pub fn signature(&self) -> &FuncSig {
+        &self.signature
+    }
+
+    pub fn call(&self, args: &[WValue]) -> Result<Vec<WValue>, String> {
+        crate::js_log(&format!(
+            "called DynFunc::call name=({}) with n args {}",
+            self.name,
+            args.len()
+        ));
+
+        let result = serde_json::ser::to_string(args);
+        if let Err(e) = result {
+            js_log(&format!("cannot serialize: {}", e));
+            return Err("cannot serialize".to_string());
+        }
+
+        let args = result.unwrap();
+        let output = INSTANCE
+            .with(|instance| call_export(instance.borrow().as_ref().unwrap(), &self.name, &args));
+
+        js_log(&format!("DynFunc::Call got result json {}", output));
+
+        let value = serde_json::de::from_str::<serde_json::Value>(&output);
+        match value {
+            Ok(serde_json::Value::Array(values)) => {
+                let values = values
+                    .iter()
+                    .map(|value| WValue::I32(value.as_i64().unwrap() as i32))
+                    .collect::<Vec<_>>();
+                Ok(values)
+            }
+            _ => {
+                js_log("invalid_json got");
+                Err("invalid json got".to_string())
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Export {
+    Memory,
+    Function(ProcessedExport),
+}
+
+impl Export {
+    pub fn name(&self) -> String {
+        match self {
+            Self::Memory => "memory".to_string(),
+            Self::Function(func) => func.name.clone(),
+        }
+    }
+}
+
+pub struct Exports {
+    exports: Vec<Export>,
+    module_name: String,
+}
+
+impl Exports {
+    pub fn new(mit: &MITInterfaces, module_name: String) -> Self {
+        let mut exports = mit
+            .exports()
+            .filter_map(|export| {
+                crate::js_log(&format!(
+                    "processing export {} {}",
+                    export.name, export.function_type
+                ));
+                let fn_type = mit.type_by_idx(export.function_type).unwrap();
+                crate::js_log(&format!("got type {}", fn_type.to_string()));
+                if let wasmer_it::ast::Type::Function {
+                    arguments,
+                    output_types,
+                } = fn_type
+                {
+                    let mut arg_types =
+                        itypes_args_to_wtypes(arguments.as_slice().iter().map(|arg| &arg.ty));
+                    let output_types = itypes_output_to_wtypes(output_types.iter());
+                    if export.name == "allocate" {
```

Comment:
I cannot do it, because`it-generator` crate transitively depends on `wasmer-runtime-core-fl`, which does not compile to wasm. So will just make a const.